[조형민] Week15

Vegatality opened the pull request:
## 요구사항

### 기본

  
- [x] [링크 공유 페이지] 링크 공유 페이지의 url path를 ‘/shared’에서 ‘/shared/{folderId}’로 변경했나요?

- [x] [링크 공유 페이지] 폴더의 정보는 ‘/api/folders/{folderId}’, 폴더 소유자의 정보는 ‘/api/users/{userId}’를 활용했나요?

- [x] [링크 공유 페이지] 링크 공유 페이지에서 폴더의 링크 데이터는 ‘/api/users/{userId}/links?folderId={folderId}’를 사용했나요?

- [x] [폴더 페이지] 폴더 페이지에서 유저가 access token이 없는 경우 ‘/signin’페이지로 이동하나요?

- [x] [폴더 페이지] 폴더 페이지의 url path가 ‘/folder’일 경우 폴더 목록에서 “전체” 가 선택되어 있고, ‘/folder/{folderId}’일 경우 폴더 목록에서 {folderId} 에 해당하는 폴더가 선택되어 있고 폴더에 있는 링크들을 볼 수 있나요?

- [x] [폴더 페이지] 폴더 페이지에서 현재 유저의 폴더 목록 데이터를 받아올 때 ‘/api/folders’를 활용했나요?

- [x] [폴더 페이지] 폴더 페이지에서 전체 링크 데이터를 받아올 때 ‘/api/links’, 특정 폴더의 링크를 받아올 때 ‘/api/links?folderId=1’를 활용했나요?

- [x] [상단 네비게이션] 유효한 access token이 있는 경우 ‘/api/users’로 현재 로그인한 유저 정보를 받아 상단 네비게이션 유저 프로필을 보여주나요?


### 심화

- [x] [심화] 리퀘스트 헤더에 인증 토큰을 첨부할 때 axios interceptors 또는 이와 유사한 기능을 활용 했나요?

## 주요 변경사항

- folder 페이지와 shared 페이지를 다이나믹 라우팅 처리하였습니다.
- folder 페이지에서 ServerSideProps를 제거하였습니다.
- 그 외에 shared 페이지에 가데이터를 사용하던 코드를 실제 데이터 베이스 데이터로 교체한 것을 제외하고는 크게 달라진 점이 없습니다.

## 스크린샷

![image](이미지url)

## 멘토에게

- accessToken과 refreshToken을 cookie에 둘 다 저장해서 SSR 하려다가 둘 다 쿠키에 저장하는 것은 위험한 것 같아서 accessToken을 localStorage에 저장하고 CSR로 불러오는 방식으로 결정하였습니다.
- 폴더 페이지에서 링크들을 불러올 때 반복적으로 api를 호출해 오는 문제가 있습니다. 추후 코드를 수정해야 할 것 같습니다.
- 셀프 코드 리뷰를 통해 질문 이어가겠습니다.

Vegatality committed:
feat: 검색 매칭 결과리스트 렌더링에 debounce 적용

Vegatality committed:
chore: 안 쓰는 useModal hook deprecated JSDoc 주석 추가

Vegatality committed:
refactor: 매칭 결과 리스트 debounce 적용한 공용 훅으로 분리

Vegatality committed:
chore: input field autofiill color 제거

Vegatality committed:
feat: fetch api를 axios로 전면 교체 및 axios interceptor 활용

Vegatality committed:
chore: console production에서 제거되도록 함.

Vegatality committed:
feat: 웹 스토리지 및 쿠키 상수와 유틸 함수 생성

Vegatality committed:
chore: SNS share 미션에 맞춰서 수정

Vegatality committed:
feat: Optional Catch-all Segments 사용해서 folder 다이나믹 라우팅

- Optional Catch-all Segments
https://nextjs.org/docs/pages/building-your-application/routing/dynamic-routes#optional-catch-all-segments

Vegatality committed:
feat: shared 페이지 다이나믹 라우팅 적용

미션에 맞춰서 다이나믹 라우팅 적용

wonsss commented on utils/logger/logOnDev.ts:
```diff
@@ -0,0 +1,5 @@
+export const logOnDev = (message: string) => {
+  if (process.env.NODE_ENV === 'development') {
+    console.log(message);
+  }
+};
```
개발환경에서만 찍히는 콘솔로그 함수군요! 유용하겠어요. 👍

wonsss commented on api/instance/axiosToken.ts:
```diff
@@ -0,0 +1,33 @@
+  timeoutErrorMessage: 'Request timeout, please try again later',
+});
+
+axiosToken.interceptors.request.use(
+  (config) => {
+    const { method, url } = config;
+    logOnDev(`🚀 [API] ${method?.toUpperCase()} ${url} | Request`);
+
+    if (getAccessToken()) {
+      config.headers.Authorization = `Bearer ${getAccessToken()}`;
+    }
+
+    return config;
```
axios interceptor를 통해 header에 accessToken을 잘 넣으셨네요 👍

wonsss commented on api/instance/axiosToken.ts:
```diff
@@ -0,0 +1,33 @@
+
+    return config;
+  },
+  (error: AxiosError | Error): Promise<AxiosError> => {
+    if (isAxiosError(error)) {
+      const { method, url } = error.config as InternalAxiosRequestConfig;
+      logOnDev(`☢️ [API] ${method?.toUpperCase()} ${url} | Response`);
+    } else {
+      logOnDev(`☢️ [API] | Error ${error.message} Response`);
+    }
+
+    return Promise.reject(error);
+  },
```
요청 시 에러가 발생하는 경우도 인터셉터에서 공통적으로 잘 처리해주셨네요. 개발환경에서 로그를 보면서 디버깅에 도움이 되겠어요

wonsss commented on api/instance/fetchWithGet.ts:
```diff
@@ -1,27 +1,24 @@
       headers: {
         'Content-Type': 'application/json',
       },
+      params: qs,
```
qs는 아마 쿼리스트링의 약자겠죠? 개인적으로 줄임말보다는 명시적으로 맞춤법에 맞는 단어를 사용하는 것을 선호합니다 ㅎㅎ
이 경우 인자명으로 qs보다는 params로 해도 되지 않을까 싶어요.

wonsss commented on api/instance/fetchWithPost.ts:
```diff
@@ -0,0 +1,27 @@
+    console.error('🚀 ~ fetchWithPost ~ error:', error);
+
+    throw new Error(getStringTypeError(error));
+  }
```
catch 블록 내에서 Error를 다시 throw하시는 이유가 궁금합니다 🤔 
여기서 catch하여 다시 error를 던지지 않더라도, 이 fetcher 함수의 사용처에서 try catch로 감싸면 에러 감지가 될 것 같아서요.
어떤 특별한 이유가 있으신지 궁금합니다.

wonsss commented on hooks/useMatchedLinksWithDebounce.ts:
```diff
@@ -0,0 +1,29 @@
+): T => {
+  const [matchedLinks, setMatchedLinks] = useState<T>(links);
+
+  const handleMatchedLinks = useCallback(
+    useDebounce(() => setMatchedLinks(links.filter((l) => filterMatchedDatas(l, input, matchingFields)) as T), wait),
+    [links, input],
+  );
+
+  useEffect(() => {
+    handleMatchedLinks();
```
여기서 디바운스를 거는 이유가 궁금해요 🤔

wonsss commented on pages/folder/comp/article/comp/card-container/comp/folder-link-category/FolderLinkCategory.module.css:
```diff
@@ -31,6 +31,8 @@
 
   padding: 0.6rem 1rem;
+
+  transition: all 0.3s ease-in-out;
```
트랜지션 잘 사용하셨네요. 👍

wonsss commented on pages/folder/comp/article/comp/card-container/comp/folder-link-category/FolderLinkCategory.tsx:
```diff
@@ -21,22 +21,28 @@ type TFolderLinkCategoryProps = {
-  }, [folderCategoryList, setFolderCategoryList]);
+interface FolderIdQuery extends ParsedUrlQuery {
+  folderId?: string[];
+}
```
전에 말씀드린 ParsedUrlQuery 잘 사용해주셨네요.
추가적으로 Zod 라는 런타임 유효성 체크 라이브러리로 url 쿼리파라미터도 체크할 수 있는데 참고해보세요.

wonsss commented on pages/folder/comp/article/comp/card-container/comp/folder-link-category/FolderLinkCategory.tsx:
```diff
@@ -21,22 +21,28 @@ type TFolderLinkCategoryProps = {
-    setFolderCategoryList(folderCategoryList);
-  }, [folderCategoryList, setFolderCategoryList]);
+interface FolderIdQuery extends ParsedUrlQuery {
+  folderId?: string[];
```
folderId 쿼리스트링에서 값이 string[] 로 받을 계획이 있으신가요?
그냥 string 으로 하는 게 맞을 것 같아 보입니다.

wonsss commented on pages/folder/comp/article/comp/card-container/comp/folder-link-category/FolderLinkCategory.tsx:
```diff
@@ -21,22 +21,28 @@ type TFolderLinkCategoryProps = {
+    handleFolderIdAndName({
+      folderId: folderId ? Number(folderId) : 'total',
+      folderName: folderId
+        ? (folderCategoryList.find((folderCategory) => folderCategory.id === Number(folderId[0]))?.name as string)
```
위에서 쿼리스트링의 타입 선언을 string으로 했다면, folderId[0] 가 아니라 그냥 folderId로 사용하실 수 있을 것 같아요.
그리고 코드가 좀 복잡한데, 명시적으로 이렇게 여러 줄에 나눠서 작성할 수도 있을 것 같아요. 명시적인 가독성 좋은 코드가 유지보수에 유리합니다.
```tsx
  useEffect(() => {
    const { folderId } = router.query as FolderIdQuery;
    const foundFolder = folderCategoryList.find(({ id }) => id === Number(folderId));
    const foundFolderName = foundFolder?.name ?? ''; // undefined인 경우 어떤 값 할당할 것인지 확인 필요

    handleFolderIdAndName({
      folderId: folderId ? Number(folderId) : 'total',
      folderName: folderId ? foundFolderName : '전체',
    });
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [folderCategoryList]);
```

wonsss commented on pages/folder/comp/article/comp/card-container/comp/folder-link-category/FolderLinkCategory.tsx:
```diff
@@ -46,7 +52,9 @@ const FolderLinkCategory = ({ selectedFolderId, handleFolderIdAndName }: TFolder
-                href={{ search: `user=${userId}&folder=${folder.id}` }}
+                // ? 페이지 url path 경로만 바꾸고 페이지 전환은 하지 않는 방법은 없나?
+                // href={{ pathname: `/folder${folder.id === 'total' ? '' : `/${encodeURIComponent(folder.id)}`}` }}
+                href={`/folder${folder.id === 'total' ? '' : `/${encodeURIComponent(folder.id)}`}`}
```
folder.id를 인코딩하신 이유가 궁금합니다.

wonsss commented on pages/shared/[folderId].page.tsx:
```diff
@@ -0,0 +1,113 @@
+export const getServerSideProps: GetServerSideProps<SharedPageProps> = async (context) => {
+  const { folderId } = context.query as SharedPageQuery;
+
+  try {
+    if (folderId === undefined) {
+      throw new Error('folderId is undefined');
+    }
+  } catch (error) {
+    console.error(getStringTypeError(error));
+
+    return {
+      redirect: {
+        destination: '/',
+        permanent: false,
+      },
+    };
+  }
```
아래와 동일한 효과의 코드일 것 같아요.
```tsx
if (folderId === undefined) {
    console.error('folderId is undefined');
 return {
      redirect: {
        destination: '/',
        permanent: false,
      },
    };
```

wonsss commented on pages/shared/[folderId].page.tsx:
```diff
@@ -0,0 +1,113 @@
+  } catch (error) {
+    console.error(getStringTypeError(error));
+
+    return {
+      redirect: {
+        destination: '/',
+        permanent: false,
+      },
+    };
```
만약 NotFound 페이지로 보내려고 한다면, 아래와 같이 하면 준비해두신 NotFound 페이지로 이동됩니다.
https://nextjs.org/docs/pages/api-reference/functions/get-server-side-props#notfound
```tsx
    return {
      notFound: true,
    };
```

wonsss commented on pages/shared/comp/banner/Banner.tsx:
```diff
@@ -16,12 +20,12 @@ const Banner = ({ userInfo }: { userInfo: TSampleFolderUserInfo | null }) => {
             className={cn('avatar')}
             alt='아바타'
-            src={!userInfo ? '/images/shared/shared-avatar.svg' : userInfo.owner.profileImageSource}
+            src={!image_source ? '/images/shared/shared-avatar.svg' : image_source}
```
`image_source ?? '/images/shared/shared-avatar.svg'` 도 같은 효과일 것 같네요. 둘 중에 더 가독성 좋은 것을 사용하시면 될 것 같아요.

wonsss commented on utils/local-storage/setToken.ts:
```diff
@@ -0,0 +1,20 @@
+  refreshToken?: string;
+};
+
+export const setToken = ({ accessToken, refreshToken }: Tokens) => {
+  if (accessToken) {
+    localStorage.setItem(ACCESS_TOKEN, JSON.stringify(accessToken));
+  }
+
+  if (refreshToken) {
+    const expires = new Date();
+    expires.setDate(expires.getDate() + 1);
+    setCookie(REFRESH_TOKEN, refreshToken, { expires, httpOnly: true, sameSite: 'Lax', secure: true });
+  }
+};
```
리프레시토큰까지 잘 해주셨네요!

wonsss commented on utils/local-storage/setToken.ts:
```diff
@@ -0,0 +1,20 @@
+  refreshToken?: string;
+};
+
+export const setToken = ({ accessToken, refreshToken }: Tokens) => {
+  if (accessToken) {
+    localStorage.setItem(ACCESS_TOKEN, JSON.stringify(accessToken));
+  }
+
+  if (refreshToken) {
+    const expires = new Date();
+    expires.setDate(expires.getDate() + 1);
+    setCookie(REFRESH_TOKEN, refreshToken, { expires, httpOnly: true, sameSite: 'Lax', secure: true });
```
쿠키 보안 위험을 고려하여 필요한 옵션들을 잘 넣어주셨네요.
httpOnly, sameSite: 'Lax', secure 각각의 역할이 무엇인지 간단히 알아보시면 좋을 것 같습니다.
Secure는 웹브라우저와 웹서버가 https로 통신하는 경우만 웹브라우저가 쿠키를 서버로 전송하는 옵션입니다.  HttpOnly는 자바스크립트의 document.cookie를 이용해서 쿠키에 접속하는 것을 막는 옵션입니다. 쿠키를 훔쳐가는 행위를 막기 위한 방법입니다.

wonsss replied:
토큰 저장 방식 관련하여 잘 해주신 것 같아요.
추가로 참고하실만한 아티클 공유드립니다.
https://angularindepth.com/posts/1382/localstorage-vs-cookies

wonsss reviewed (changes requested):
안녕하세요 형민님!
항상 깊게 고민하시고 좋은 코드를 잘 작성하시는 것 같아요.
이번 미션도 수고 많으셨습니다.
리뷰 반영 후 재리뷰 요청주세요. 감사합니다 😄

Vegatality committed:
refactor: fetcher함수들 리팩토링

1. 불필요한 catch문 제거
2. params에 부여되는 변수명 qs에서 params로 변경
기타. 필요없는 주석 제거

Vegatality committed:
fix: folderPage Optional Catch-all Segments에 맞춰 query 배열 형태로 명확하게 수정

Vegatality committed:
refactor: shared 페이지 리팩토링

1. [folderId] 다이나믹 라우팅 페이지 불필요한 try catch 문 리팩토링
2. banner 컴포넌트 image source null 병합 연산자로 수정.
